libplatsupport/src/mach/spacemit/timer.c

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright 2025, 10xEngineers 
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <assert.h>
+#include <errno.h>
+#include <inttypes.h>
+#include <platsupport/mach/timer.h>
+#include <stdint.h>
+#include <utils/util.h>
+
+#define MAX_TIMEOUT_NS (SPACEMIT_TIMER_MAX_TICKS * (NS_IN_S / SPACEMIT_TIMER_TICKS_PER_SECOND))
+
+void spacemit_timer_enable(spacemit_timer_t *timer)
+{
+    assert(timer);
+    assert(timer->regs);
+
+    timer->regs->ctrl |= 0x1;
+}
+
+void spacemit_timer_disable(spacemit_timer_t *timer)
+{
+    assert(timer);
+    assert(timer->regs);
+
+    timer->regs->ctrl &= 0xfffffffe;
+}
+
+void spacemit_timer_handle_irq(spacemit_timer_t *timer)
+{
+    assert(timer);
+    assert(timer->regs);
+
+    timer->value_h += 1;
+
+    /* Clear interrupt status */
+    timer->regs->status = 0x1;
+}
+
+uint64_t spacemit_timer_get_time(spacemit_timer_t *timer)
+{
+    assert(timer);
+    assert(timer->regs);
+
+    /* Read 64-bit counter safely */
+    uint32_t hi1, lo, hi2;
+    
+    do {
+        hi1 = timer->regs->cnt_hi;
+        lo = timer->regs->cnt_lo;
+        hi2 = timer->regs->cnt_hi;
+    } while (hi1 != hi2);
+
+    uint64_t value_ticks = (((uint64_t)hi1) << 32) | lo;
+    uint64_t value_h = (uint64_t)timer->value_h;
+
+    /* Include unhandled interrupt in value_h */
+    if (timer->regs->status & 0x1) {
+        value_h += 1;
+    }
+
+    value_ticks += (value_h << 32);
+
+    /* convert from ticks to nanoseconds */
+    uint64_t value_whole_seconds = value_ticks / SPACEMIT_TIMER_TICKS_PER_SECOND;
+    uint64_t value_subsecond_ticks = value_ticks % SPACEMIT_TIMER_TICKS_PER_SECOND;
+    uint64_t value_subsecond_ns =
+        (value_subsecond_ticks * NS_IN_S) / SPACEMIT_TIMER_TICKS_PER_SECOND;
+    uint64_t value_ns = value_whole_seconds * NS_IN_S + value_subsecond_ns;
+
+    return value_ns;
+}
+
+void spacemit_timer_reset(spacemit_timer_t *timer)
+{
+    assert(timer);
+    assert(timer->regs);
+
+    timer->regs->ctrl |= SPACEMIT_TIMER_MODE_FREE_RUNNING;
+    timer->regs->match_lo = 0xffffffff;
+    timer->regs->match_hi = 0xffffffff;
+    timer->value_h = 0;
+}
+
+int spacemit_timer_set_timeout(spacemit_timer_t *timer, uint64_t ns, bool is_periodic)
+{
+    spacemit_timer_disable(timer);
+    timer->value_h = 0;
+
+    if (is_periodic) {
+        timer->regs->ctrl = SPACEMIT_TIMER_MODE_FREE_RUNNING;
+    } else {
+        timer->regs->ctrl = SPACEMIT_TIMER_MODE_USER_DEFINED;
+    }
+
+    uint64_t ticks_whole_seconds = (ns / NS_IN_S) * SPACEMIT_TIMER_TICKS_PER_SECOND;
+    uint64_t ticks_remainder = (ns % NS_IN_S) * SPACEMIT_TIMER_TICKS_PER_SECOND / NS_IN_S;
+    uint64_t num_ticks = ticks_whole_seconds + ticks_remainder;
+
+    /* Read current counter value */
+    uint32_t hi1, lo, hi2;
+    do {
+        hi1 = timer->regs->cnt_hi;
+        lo = timer->regs->cnt_lo;
+        hi2 = timer->regs->cnt_hi;
+    } while (hi1 != hi2);
+
+    uint64_t current_ticks = (((uint64_t)hi1) << 32) | lo;
```

Comment:
Better to create a small helper function that does this and use it both here and in `spacemit_timer_get_time()`.